### PR TITLE
CI: switch one macOS CI job from distutils to meson

### PR DESCRIPTION
**Pull request by rgommers**
This should be a little faster, and test something different - the two macOS jobs that we have now were basically identical. This job has been stable for a little while, so it's good to have it in the main repo now. 

Also important to point out: it's our first CI job using `conda-forge`, which is a nice addition. The rough comparison of timings:
- `setup.py` based job: 39-48 min
- `meson` based job: 29-41 min

The meson job takes 3-4 extra for installing with `mamba` (more needs to be downloaded, like Python itself), but builds a few minutes faster on a cold cache (and >5 min for rebuilds due to ccache), and the tests run >5 min faster as well because it's optimized for 2 cores, unlike the `runtests.py` based CI job.

### CODE REVIEW

**Comment by rgommers** [comment]
Ugh, all I did was change Python 3.9 to 3.10 to match the existing CI job, and now I see:
```
ld: warning: object file (scipy/special/_ufuncs.cpython-310-darwin.so.p/meson-generated__ufuncs.c.o)
was built for newer macOS version (11.0) than being linked (10.16)
```
The image provides macOS `11.6.2`.

**Comment by rgommers** [comment]
And just to prove me wrong, the `setup.py` based job had a really fast run: 36 min.

**Comment by andyfaff** [comment]
Thanks for this Ralf, but I'd say building via conda-forge is not necessarily a plus point. Wouldn't the conda-forge feedstock be the canary in the mine for indicating problems with building via conda?

**Comment by rgommers** [comment]
> Thanks for this Ralf, but I'd say building via conda-forge is not necessarily a plus point. Wouldn't the conda-forge feedstock be the canary in the mine for indicating problems with building via conda?

No it wouldn't, it's normally only triggered by a release we make on PyPI, or by someone doing manual testing on the feedstock.

The way any issues would otherwise show up is on someone's local dev environment. I think this is a useful change.

**Comment by rgommers** [comment]
Also, conda-forge is our recommended development environment (http://scipy.github.io/devdocs/dev/contributor/quickerstart_conda.html#quickerstart-conda), since it's the easiest to give complete and reproducible instructions for. So having one job to test it is good - we should perhaps add more in the future. Right now our CI configs are not optimal, with multiple jobs often testing essentially the same thing.

**Comment by rgommers** [comment]
> `was built for newer macOS version (11.0) than being linked (10.16)`

That was caused by running on a macOS 11 image somehow, and fixed by setting `export MACOSX_SDK_VERSION=10.9`.

This is good to go now.

File: .github/workflows/macos_meson.yml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+env:
+  INSTALLDIR: "installdir"
+  CCACHE_DIR: "${{ github.workspace }}/.ccache"
+
+jobs:
+  test_meson:
+    name: Meson build
+    # If using act to run CI locally the github object does not exist and the usual skipping should not be enforced
+    if: "github.repository != '' || github.repository == 'scipy/scipy' && !contains(github.event.head_commit.message, '[ci skip]') && !contains(github.event.head_commit.message, '[skip ci]') && !contains(github.event.head_commit.message, '[skip github]')"
```
Comment (andyfaff) [inline .github/workflows/macos_meson.yml:21]: I find the logic of this line hard to follow. I think the gist is to run the tests if:

- if you're running locally, no matter that the commit message is.
- if you're running on github, skip the test if the commit message contains skip-ci, etc.

File: .github/workflows/macos_meson.yml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+          ${{ github.workflow }}-${{ matrix.python-version }}-ccache-macos-
+
+    - name: Setup Conda
+      uses: conda-incubator/setup-miniconda@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+        mamba-version: "*"
+        channels: conda-forge
+        channel-priority: true
+        activate-environment: scipy-meson
```
Comment (andyfaff) [inline .github/workflows/macos_meson.yml:75]: I assume that this activates a scipy-meson environment. How can this be activated if an environment is not created yet? Does L90 actually create the environment?

Comment (andyfaff) [reply .github/workflows/macos_meson.yml:75]: Is this line needed, given that L109 seems to activate the environment as well.

File: .github/workflows/macos_meson.yml (hunk 1 above)
Comment (rgommers) [reply .github/workflows/macos_meson.yml:21]: There are 3 parts:
- allow running locally (with Act)
- disable running on forks by default (that's the `github.repository == 'scipy/scipy`)
- add logic to do the right thing for `[skip github]` and `[skip ci]` in commit messages

I agree this is hard to parse. How about we add a comment explaining the above to all CI jobs?

File: .github/workflows/macos_meson.yml (hunk 2 above)
Comment (rgommers) [reply .github/workflows/macos_meson.yml:75]: Yes, this follows the docs in https://github.com/marketplace/actions/setup-miniconda#caching-environments

Comment (andyfaff) [reply .github/workflows/macos_meson.yml:75]: The documentation on that link is unusually clear.

File: .github/workflows/macos_meson.yml (hunk 1 above)
Comment (andyfaff) [reply .github/workflows/macos_meson.yml:21]: I think I got confused by the associative nature of the booleans, as well as the quotes. I think it's probably mergeable as is. On re-reading it seems a lot clearer.

**Comment by andyfaff** [comment]
Thanks Ralf.
